include/ginkgo/core/preconditioner/ilu.hpp

```diff
@@ -0,0 +1,380 @@
+    {
+        ensure_cache_support_for(b);
+        if (!ReverseApply) {
+            l_solver_->apply(b, cache_.intermediate.get());
+            u_solver_->apply(alpha, cache_.intermediate.get(), beta, x);
+        } else {
+            u_solver_->apply(b, cache_.intermediate.get());
+            l_solver_->apply(alpha, cache_.intermediate.get(), beta, x);
+        }
+    }
+
+    explicit Ilu(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Ilu>(std::move(exec))
+    {}
+
+    explicit Ilu(const Factory *factory, LuArgs lu_args)
+        : EnableLinOp<Ilu>(factory->get_executor(),
+                           lu_args.get_solver_size(ReverseApply)),
+          parameters_{factory->get_parameters()},
+          l_factor_{std::move(lu_args.l_factor)},
+          u_factor_{std::move(lu_args.u_factor)}
+    {
+        GKO_ASSERT_EQUAL_DIMENSIONS(l_factor_, u_factor_);
+
+        auto exec = this->get_executor();
+
+        // If no factories are provided, generate default ones
+        if (!parameters_.l_solver_factory) {
+            l_solver_ = generate_default_solver<l_solver_type>(exec, l_factor_);
+        } else {
+            l_solver_ = parameters_.l_solver_factory->generate(l_factor_);
+        }
+        if (!parameters_.u_solver_factory) {
+            u_solver_ = generate_default_solver<u_solver_type>(exec, u_factor_);
+        } else {
+            u_solver_ = parameters_.u_solver_factory->generate(u_factor_);
+        }
+    }
+
+    void ensure_cache_support_for(const LinOp *b) const
+    {
+        dim<2> expected_size =
+            ReverseApply ? dim<2>{u_solver_->get_size()[0], b->get_size()[1]}
+                         : dim<2>{l_solver_->get_size()[0], b->get_size()[1]};
+        if (cache_.intermediate == nullptr ||
+            cache_.intermediate->get_size() != expected_size) {
+            cache_.intermediate = matrix::Dense<value_type>::create(
+                this->get_executor(), expected_size);
+        }
+    }
+
+    /**
+     * @internal  Looks at the build() method to determine the type of the
+     *            factory.
+     */
+    template <typename T>
+    using factory_type_t = decltype(T::build());
+
+    // Parameter type of function `with_criteria`.
+    using with_criteria_param_type =
+        std::shared_ptr<const stop::CriterionFactory>;
+
+    /**
+     * Helper structure to test if the Factory of SolverType has a function
+     * `with_criteria`.
+     *
+     * Contains a constexpr boolean `value`, which is true if the Factory class
+     * of SolverType has a `with_criteria`, and false otherwise.
+     *
+     * @tparam SolverType   Solver to test if its factory has a with_criteria
+     *                      function.
+     *
+     */
+    template <typename SolverType, typename = void>
+    struct has_with_criteria : std::false_type {};
+
+    /**
+     * @copydoc has_with_criteria
+     *
+     * @internal  The second template parameter (which uses SFINAE) must match
```

Comment:
what is SFINAE

Reply:
[SFINAE: Substitution Failure Is Not An Error](https://en.wikipedia.org/wiki/Substitution_failure_is_not_an_error), which means that if there is an invalid substitution of a template parameter, it does not count as a compiler error. It will act as if this function / structure / class does not exist, and continues the search for the best fit substitution (skipping every invalid substitution).